resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2024 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.util.internal.MathUtil;
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Random;
+
+/**
+ * Special data-structure that will allow to retrieve the next query id to use, while still guarantee some sort
+ * of randomness.
+ * The query id will be between 0 (inclusive) and 65535 (inclusive) as defined by the RFC.
+ */
+final class DnsQueryIdSpace {
+    private static final int MAX_ID = 65535;
+
+    // Holds all possible ids which are stored as unsigned shorts
+    private final short[] ids = new short[MAX_ID + 1];
```

Comment:
Perhaps for a future optimization we could make this growable and start small. That should keep it roughly on the same footprint as the `Map` that holds the state objects.

Reply:
Scratch this, sorry: it's actually the opposite: the size of this is at it's max when no id's have been issues which is the exact opposite of the `Map` that holds the state objects.

Reply:
Yeah... I was also thinking about ways how to grow / shrink but all I came up with would result in less randomness which I think is not good